src/poly/commitment.rs

```diff
@@ -154,6 +217,95 @@ impl<C: CurveAffine> Params<C> {
     }
 }
 
+/// A guard returned by the verifier
+#[derive(Debug)]
+pub struct Guard<'a, C: CurveAffine> {
```

Comment:
As I mentioned in the review comment above, it could actually also contain the `challenges_sq`/`allinv` values produced by the verifier so that they don't have to be recomputed in `use_s`.

Reply:
`Guard` now looks like: 

```
pub struct Guard<'a, C: CurveAffine> {
    msm: &'a mut MSM<C>,
    neg_z1: C::Scalar,
    allinv: C::Scalar,
    challenges_sq: Vec<C::Scalar>,
}
```